### Question 1

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.feature_extraction.text import TfidfVectorizer
import random
import tensorflow as tf

In [2]:
#Firstly we are going to train our model through the url from the manual
import pandas as pd

DATASET_URL = 'https://gist.githubusercontent.com/amitness/0a2ddbcb61c34eab04bad5a17fd8c86b/raw/66ad13dfac4bd1201e09726677dd8ba8048bb8af/clickbait.csv'
data = pd.read_csv(DATASET_URL)
data.head(5)

title  label
0  15 Highly Important Questions About Adulthood,...      1
1  250 Nuns Just Cycled All The Way From Kathmand...      1
2  Australian comedians "could have been shot" du...      0
3  Lycos launches screensaver to increase spammer...      0
4  Fußball-Bundesliga 2008–09: Goalkeeper Butt si...      0

In [3]:
#train the model  and split it 80/20 ratio train set/test set
X = list(data.title.values)
y = list(data.label.values)# the labels we want to predict --> Y
labels = ['not clickbait', 'clickbait']

X_train_str, X_test_str, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

In [4]:
#create the vector of occurance, for every word
cv = CountVectorizer() # this initializes the CountVectorizer 

cv.fit(X_train_str) # create the vocabulary

X_train = cv.transform(X_train_str)
X_test = cv.transform(X_test_str)

In [5]:
#print(X_train.toarray()[0])

In [6]:
#i wanted to represent the vocbulary as a dataframe to visualize the result
vocabulary = cv.get_feature_names()
vectorized_texts = pd.DataFrame(X_train.toarray(), columns=vocabulary)
vectorized_texts.head(5)

00  000  000th  00s  04  05  08  08m  09  10  ...  zuckerberg  zuma  \
0   0    0      0    0   0   0   0    0   0   0  ...           0     0   
1   0    0      0    0   0   0   0    0   0   0  ...           0     0   
2   0    0      0    0   0   0   0    0   0   0  ...           0     0   
3   0    0      0    0   0   0   0    0   0   0  ...           0     0   
4   0    0      0    0   0   0   0    0   0   0  ...           0     0   

   zurawski  zurich  złoty  íngrid  íslands  île  śrī  šibenik  
0         0       0      0       0        0    0    0        0  
1         0       0      0       0        0    0    0        0  
2         0       0      0       0        0    0    0        0  
3         0       0      0       0        0    0    0        0  
4         0       0      0       0        0    0    0        0  

[5 rows x 20660 columns]

In [7]:
#train the model with Logistic Regression
lr = LogisticRegression(solver='lbfgs',max_iter=1000)
lr.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=1000,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [8]:
#load my dataset for the exercise
data2 = pd.read_pickle("framing.p")

In [9]:
del data2['label'] #i know that i could just name my new column otherwise but i like the name labels and for this example the label didint do anything

In [10]:
data2.head()

tweet_id                date       user party    state  chamber  \
0  1325914751495499776 2020-11-09 21:34:45  SenShelby     R  Alabama  Senator   
1  1294021087118987264 2020-08-13 21:20:43  SenShelby     R  Alabama  Senator   
2  1323340848130609156 2020-11-02 19:06:59  DougJones     D  Alabama  Senator   
3  1323004075831709698 2020-11-01 20:48:46  DougJones     D  Alabama  Senator   
4  1322567531320717314 2020-10-31 15:54:06  DougJones     D  Alabama  Senator   

                                               tweet     news_mention  \
0  ICYMI – @BusinessInsider declared #Huntsville ...  businessinsider   
1  Great news! Today @mazda_toyota announced an a...                    
2  He’s already quitting on the folks of Alabama ...                    
3  I know you guys are getting bombarded with fun...                    
4  Well looky here folks, his own players don’t t...                    

                                       url_reference                   netloc  \
0  https://www.businessinsider.com/personal-finan...  www.businessinsider.com   
1  https://pressroom.toyota.com/mazda-and-toyota-...     pressroom.toyota.com   
2  https://apnews.com/article/c73f0dfe8008ebaf85e...               apnews.com   
3  https://secure.actblue.com/donate/djfs-close?r...       secure.actblue.com   
4  https://slate.com/culture/2020/10/tommy-tuberv...                slate.com   

                                               title  \
0  The 10 best US cities to move to if you want t...   
1  Mazda and Toyota Further Commitment to U.S. Ma...   
2  Tuberville, Jones fight for Senate seat in Ala...   
3                                       I just gave!   
4  What Tommy Tuberville’s Former Auburn Players ...   

                                         description  
0  The best US cities to move to if you want to s...  
1  HUNTSVILLE, Ala., (Aug. 13, 2020) – Today, Maz...  
2  GARDENDALE, Ala. (AP) — U.S. Sen. Doug Jones, ...  
3                         Join us! Contribute today.  
4                       "All I could think is, why?"

In [11]:
new_example =list(data2['title'])

In [12]:
#train my new dataset to the existing model thati created before and add the predictions to a constant
new_example_bow = cv.transform(new_example)# icould use more examples like tf-idf or Bert my choice was 
predictions = lr.predict(new_example_bow)
a = []
for i, prediction in enumerate(predictions):   # i use that for loop so i dont have to rename tha labes from 0 or 1
    "{}: {}".format(new_example[i], labels[prediction])
    a.append(labels[prediction])

In [13]:
data2['labels'] = a

In [14]:
data2.head(5)

tweet_id                date       user party    state  chamber  \
0  1325914751495499776 2020-11-09 21:34:45  SenShelby     R  Alabama  Senator   
1  1294021087118987264 2020-08-13 21:20:43  SenShelby     R  Alabama  Senator   
2  1323340848130609156 2020-11-02 19:06:59  DougJones     D  Alabama  Senator   
3  1323004075831709698 2020-11-01 20:48:46  DougJones     D  Alabama  Senator   
4  1322567531320717314 2020-10-31 15:54:06  DougJones     D  Alabama  Senator   

                                               tweet     news_mention  \
0  ICYMI – @BusinessInsider declared #Huntsville ...  businessinsider   
1  Great news! Today @mazda_toyota announced an a...                    
2  He’s already quitting on the folks of Alabama ...                    
3  I know you guys are getting bombarded with fun...                    
4  Well looky here folks, his own players don’t t...                    

                                       url_reference                   netloc  \
0  https://www.businessinsider.com/personal-finan...  www.businessinsider.com   
1  https://pressroom.toyota.com/mazda-and-toyota-...     pressroom.toyota.com   
2  https://apnews.com/article/c73f0dfe8008ebaf85e...               apnews.com   
3  https://secure.actblue.com/donate/djfs-close?r...       secure.actblue.com   
4  https://slate.com/culture/2020/10/tommy-tuberv...                slate.com   

                                               title  \
0  The 10 best US cities to move to if you want t...   
1  Mazda and Toyota Further Commitment to U.S. Ma...   
2  Tuberville, Jones fight for Senate seat in Ala...   
3                                       I just gave!   
4  What Tommy Tuberville’s Former Auburn Players ...   

                                         description         labels  
0  The best US cities to move to if you want to s...      clickbait  
1  HUNTSVILLE, Ala., (Aug. 13, 2020) – Today, Maz...  not clickbait  
2  GARDENDALE, Ala. (AP) — U.S. Sen. Doug Jones, ...  not clickbait  
3                         Join us! Contribute today.      clickbait  
4                       "All I could think is, why?"  not clickbait

Now we will count how many times democrats and republicans use clickbait titles and compare them

In [15]:
#we create 2 new attributes so its easy to count the numbers of each other party labels
data2_rep = data2[data2['party']=='R']
data2_dem = data2[data2['party']=='D']
print("Republicans are :",len(data2_rep))
print("Democrats are :",len(data2_dem))


Republicans are : 7771
Democrats are : 15564


In [16]:
#thats how many times repablicans used clickbait labales
rep_times = data2_rep[data2_rep.labels == 'clickbait']['labels'].count()
print(rep_times)

1058


In [17]:
#thats how many times democrats used clickbait labales
dem_times = data2_dem[data2_dem.labels == 'clickbait']['labels'].count()
print(dem_times)

2830


And if we want the proportion of both partys:

In [18]:
#now for the repablicans
proportion_rep_click = rep_times/len(data2_rep)
print(proportion_rep_click)

0.1361472140007721


In [19]:
#and now for the democrats
proportion_dem_click = dem_times/len(data2_dem)
print(proportion_dem_click)


0.18182986378822924


So in conclusion the democrats refer to an article with a clickbait title 2830 times while the republicans only 1058 even the difference in the sample is big democrats are way more in this sample the proportion is still bigger for democrats rather than republicans.
After viewing the clickbaited onces i saw that most of them where not making any sence and they where either rubbish or advertisments,also they seem to have some actual "conversation" like "i know that you boomed with advertisments" so they most are in 1st person as far as i reviwed them.